### PR TITLE
Add MODULATE builtin to canvas item shaders (3.2)

### PR DESCRIPTION
**Pull request by RandomShaper**
This adds a new builtin for canvas item shaders,  `in color MODULATE`. Its value is the final modulate color that would be applied to the item.

By default, modulate color is applied after the user fragment shader code has been run. This addition allows you to use the modulate yourself however you want. Therefore, if you use `MODULATE`, the default behavior won't happen so you are responsible of multiplying at the end (that would be replicating the default behavior, so quite pointless) or doing something fancier with it.

**My use case:** I have an almost finished game in Godot 2.1 that uses canvas item shaders to create rich looks by using normal maps and applying the modulate color in 3D-like ways, so to speak. After porting it to 3.2, it wasn't possible anymore to do that. This PR re-adds that ability.

There will be a separate PR for _master_, in case the shading model for 4.0 makes it necessary.

### CODE REVIEW

**Comment by lawnjelly** [comment]
I've spoken to RandomShaper and it requires a slight tweak of the batching to work with it (because we bake the final modulate and vertex colors together in some batches), but it looks quite straightforward to get it working, we can just detect the MODULATE keyword and disable translating to colored verts for that item. :+1: 

Just as soon as I can remember where @clayjohn showed me how to detect keywords in the shaders lol :grin: ...

File: scene/resources/visual_shader.cpp [hunk 1]
```diff
@@ -1536,6 +1536,7 @@ const VisualShaderNodeInput::Port VisualShaderNodeInput::ports[] = {
 	{ Shader::MODE_CANVAS_ITEM, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_VECTOR, "uv", "vec3(UV, 0.0)" },
 	{ Shader::MODE_CANVAS_ITEM, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_VECTOR, "color", "COLOR.rgb" },
 	{ Shader::MODE_CANVAS_ITEM, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_SCALAR, "alpha", "COLOR.a" },
+	{ Shader::MODE_CANVAS_ITEM, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_VECTOR, "modulate", "MODULATE.rgba" },
```
Comment (arkology) [inline scene/resources/visual_shader.cpp:1539]: As I know, we don't have vec4 in visual shaders

Comment (Chaosus) [reply scene/resources/visual_shader.cpp:1539]: we can introduce modulate_color and modulate_alpha...

Comment (RandomShaper) [reply scene/resources/visual_shader.cpp]: Indeed. Fixed.

**Comment by RandomShaper** [comment]
I've updated the PR with a change: `MODULATE` can be read from the vertex shader as well. That's useful if you want to use it to generate the `COLOR` somehow from it. That way you don't need to do it per every fragment. Or if you just want to pre-modulate the color, just just have to do it per vertex.

The same semantics of inhibiting the default post-modulate behavior still hold.

@lawnjelly, I guess this doesn't change anything for our plans. :)

**Comment by akien-mga** [comment]
Thanks!

Don't forget to do the matching PR for 4.0 :)

**Comment by RandomShaper** [comment]
It seems a PR for 4.0 is not needed now. How canvas item shaders will work there still has some open questions and maybe in the end we don't need the distinction between `COLOR` and `MODULATE` anymore in 4.0 and a simpler model can be used there. `MODULATE` was needed in 3.2 to more control without breaking compatibility.

What I'll do is documenting `MODULATE` in the docs repo.

**Comment by lawnjelly** [comment]
I was thinking about this last night, and there may be some other effects with color in a custom shader, I need to examine this in detail. It may be worth either swapping to / or adding a vertex format including color and final modulate separately for this purpose. It's not a big deal, it's just the vertex format will be larger, so there will be some small drop in efficiency doing that. The shader itself will be doing the same work.

**Pull request merged**